test/testers/rust-tester/src/pdre_api/crypto.rs

```diff
@@ -0,0 +1,179 @@
+        output
+    );
+}
+
+pub fn test_ed25519() {
+    let mut api = CryptoApi::new();
+
+    // Generate key pair
+    let keystore = String::from("dumy");
+    let mut pubkey1 = [0; 32]; // will get generated
+
+    api.rtm_ext_ed25519_generate(keystore.as_bytes(), &[], &mut pubkey1);
+
+    // Sign a message
+    let msg = String::from("This is a message");
+    let mut signature = [0; 64]; // will get generated
+
+    let res = api.rtm_ext_ed25519_sign(
+        keystore.as_bytes(),
+        &pubkey1,
+        msg.as_bytes(),
+        &mut signature,
+    );
+    assert_eq!(res, 0);
+
+    // Verify message
+    let verify = api.rtm_ext_ed25519_verify(msg.as_bytes(), &signature, &pubkey1);
+    assert_eq!(verify, 0);
+
+    // Generate new key pair for listing
+    let mut pubkey2 = [0; 32]; // will get generated
+    api.rtm_ext_ed25519_generate(keystore.as_bytes(), &[], &mut pubkey2);
+
+    // Get all public keys
+    let mut result_len: u32 = 0;
+    let all_pubkeys = api.rtm_ext_ed25519_public_keys(keystore.as_bytes(), &mut result_len);
+    //assert_eq!(result_len, 65); // Why 65 and not 64?
+
+    println!("> testing ed25519 keys");
+    println!("  - public key 1: {}", hex::encode(pubkey1));
+    println!("  - message: {}", msg);
+    println!("  - signature: {}", hex::encode(&signature[..]));
+    print!("  - signature valid?: ");
+    if verify == 0 {
+        println!("YES");
+    } else {
+        println!("NO");
+    }
+    println!("  - public key 2: {}", hex::encode(pubkey2));
+    println!("  - all public keys : {}", hex::encode(&all_pubkeys[1..])); // TODO; should be [..]
+}
+
+pub fn test_sr25519() {
+    let mut api = CryptoApi::new();
+
+    // Generate key pair
+    let keystore = String::from("dumy");
+    let mut pubkey1 = [0; 32]; // will get generated
+
+    api.rtm_ext_sr25519_generate(keystore.as_bytes(), &[], &mut pubkey1);
+
+    // Sign a message
+    let msg = String::from("This is a message");
+    let mut signature = [0; 64]; // will get generated
+
+    let res = api.rtm_ext_sr25519_sign(
+        keystore.as_bytes(),
+        &pubkey1,
+        msg.as_bytes(),
+        &mut signature,
+    );
+    assert_eq!(res, 0);
+
+    let verify = api.rtm_ext_sr25519_verify(msg.as_bytes(), &signature, &pubkey1);
+    assert_eq!(verify, 0);
+
+    // Generate new key pair for listing
+    let mut pubkey2 = [0; 32]; // will get generated
+    api.rtm_ext_sr25519_generate(keystore.as_bytes(), &[], &mut pubkey2);
+
+    // Get all public keys
+    let mut result_len: u32 = 0;
+    let all_pubkeys = api.rtm_ext_sr25519_public_keys(keystore.as_bytes(), &mut result_len);
+    assert_eq!(result_len, 65);
+
+    println!("> testing sr25519 keys");
+    println!("  - public key 1: {}", hex::encode(pubkey1));
+    println!("  - message: {}", msg);
+    println!("  - signature: {}", hex::encode(&signature[..]));
+    print!("  - signature valid?: ");
+    if verify == 0 {
+        println!("YES");
```

Comment:
"Good signature" is a better output upper or lower case.

Reply:
Done!